### PR TITLE
[quant] Finishing refactor for quantization test files

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37366 [quant] Finishing refactor for quantization test files**

Summary:
- we can put both fake quant module and observer module tests in the test_workflow_module.py
- added test_quantized_functional.py
- moved tests in test_numerics.py to test_quantize.py and removed test_numerics.py

Test Plan:
python test/test_quantization.py

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D21282198](https://our.internmc.facebook.com/intern/diff/D21282198)

### CODE REVIEW

File: test/quantization/test_fake_quant.py [hunk 1]
```diff
@@ -1,311 +0,0 @@
...
-
-NP_RANDOM_SEED = 19
-tolerance = 1e-6
-
-class TestFakeQuantizePerTensor(TestCase):
-
-    @given(device=st.sampled_from(['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']),
-           X=hu.tensor(shapes=hu.array_shapes(1, 5,),
-                       qparams=hu.qparams(dtypes=torch.quint8)))
-    def test_forward_per_tensor(self, device, X):
```
Comment (raghuramank100) [inline test/quantization/test_fake_quant.py:71]: Some of the fake-quant tests are actually tests of the aten fake-quant op. These op tests should be ideally moved to test_quantized_op.py

File: test/quantization/test_fake_quant.py [hunk 2]
```diff
@@ -1,311 +0,0 @@
...
-        Y = torch.dequantize(torch.quantize_per_tensor(X.cpu(), scale, zero_point, torch_type))
-        Y_prime = torch.fake_quantize_per_tensor_affine(
-            X, scale, zero_point, quant_min, quant_max)
-        np.testing.assert_allclose(Y, Y_prime.cpu(), rtol=tolerance, atol=tolerance)
-
-    @given(device=st.sampled_from(['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']),
-           X=hu.tensor(shapes=hu.array_shapes(1, 5,),
-                       qparams=hu.qparams(dtypes=[torch.quint8])),
-           )
-    def test_fq_module(self, device, X):
```
Comment (raghuramank100) [inline test/quantization/test_fake_quant.py:132]: This is a module level test , should be ideally in test_quantized_module.py

File: test/quantization/test_fake_quant.py [hunk 3]
```diff
@@ -1,311 +0,0 @@
...
-        Y = _fake_quantize_per_tensor_affine_reference(X, fq_module.scale, fq_module.zero_point, quant_min, quant_max)
-        np.testing.assert_allclose(Y.cpu().detach().numpy(), Y_prime.cpu().detach().numpy(), rtol=tolerance, atol=tolerance)
-
-        # Test backward
-        dout = torch.rand(X.shape, dtype=torch.float, device=device)
-        Y_prime.backward(dout)
-        dX = _fake_quantize_per_tensor_affine_grad_reference(dout, X, fq_module.scale, fq_module.zero_point, quant_min, quant_max)
-        np.testing.assert_allclose(dX.cpu().numpy(), X.grad.cpu().detach().numpy(), rtol=tolerance, atol=tolerance)
-
-    def test_fq_serializable(self):
```
Comment (raghuramank100) [inline test/quantization/test_fake_quant.py:153]: This one too

File: test/quantization/test_fake_quant.py [hunk 4]
```diff
@@ -1,311 +0,0 @@
...
-        b.seek(0)
-        loaded_dict = torch.load(b)
-        loaded_fq_module = FakeQuantize(observer, quant_min, quant_max)
-        loaded_fq_module.load_state_dict(loaded_dict)
-        for key in state_dict:
-            self.assertEqual(state_dict[key], loaded_fq_module.state_dict()[key])
-
-        self.assertEqual(loaded_fq_module.calculate_qparams(), fq_module.calculate_qparams())
-
-    def test_fake_quant_control(self):
```
Comment (raghuramank100) [inline test/quantization/test_fake_quant.py:174]: Ditto for this one, this is also a module level test

File: test/quantization/test_fake_quant.py [hunk 5]
```diff
@@ -1,311 +0,0 @@
...
-        self.assertNotEqual(fq_module.zero_point, zero_point)
-
-
-
-class TestFakeQuantizePerChannel(TestCase):
-
-    @given(device=st.sampled_from(['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']),
-           X=hu.per_channel_tensor(shapes=hu.array_shapes(1, 5,),
-           qparams=hu.qparams(dtypes=torch.quint8)))
-    def test_forward_per_channel(self, device, X):
```
Comment (raghuramank100) [inline test/quantization/test_fake_quant.py:210]: This is an op level test, should move to test_quantized_op.py

File: test/quantization/test_fake_quant.py [hunk 6]
```diff
@@ -1,311 +0,0 @@
...
-        zero_point = torch.tensor(zero_point).to(dtype=torch.int64, device=device)
-        Y = _fake_quantize_per_channel_affine_reference(X.cpu(), scale.cpu(), zero_point.cpu(), axis, quant_min, quant_max)
-        Y_prime = torch.fake_quantize_per_channel_affine(
-            X, scale, zero_point, axis, quant_min, quant_max)
-        np.testing.assert_allclose(Y, Y_prime.cpu(), rtol=tolerance, atol=tolerance)
-
-    @given(device=st.sampled_from(['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']),
-           X=hu.per_channel_tensor(shapes=hu.array_shapes(1, 5,),
-           qparams=hu.qparams(dtypes=torch.quint8)))
-    def test_backward_per_channel(self, device, X):
```
Comment (raghuramank100) [inline test/quantization/test_fake_quant.py:229]: Same here, this is an op level test

File: test/quantization/test_fake_quant.py [hunk 7]
```diff
@@ -1,311 +0,0 @@
...
-        dX = _fake_quantize_per_channel_affine_grad_reference(
-            dout, X, scale, zero_point, axis, quant_min, quant_max)
-        Y_prime.backward(dout)
-        np.testing.assert_allclose(dX.cpu().detach().numpy(), X.grad.cpu().detach().numpy(), rtol=tolerance, atol=tolerance)
-
-    @given(device=st.sampled_from(['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']),
-           X=hu.per_channel_tensor(shapes=hu.array_shapes(1, 5,),
-           qparams=hu.qparams(dtypes=torch.quint8)))
-    @unittest.skip("temporarily disable the test")
-    def test_numerical_consistency_per_channel(self, device, X):
```
Comment (raghuramank100) [inline test/quantization/test_fake_quant.py:253]: This is also an op level test

File: test/quantization/test_fake_quant.py [hunk 8]
```diff
@@ -1,311 +0,0 @@
...
-        Y = _fake_quantize_per_tensor_affine_reference(X.cpu(), scale, zero_point, quant_min, quant_max)
-        Y_prime = torch.fake_quantize_per_tensor_affine(
-            X, scale, zero_point, quant_min, quant_max)
-        np.testing.assert_allclose(Y, Y_prime.cpu(), rtol=tolerance, atol=tolerance)
-
-    @given(device=st.sampled_from(['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']),
-           X=hu.tensor(shapes=hu.array_shapes(1, 5,),
-                       qparams=hu.qparams(dtypes=torch.quint8)))
-    @unittest.skip("temporarily disable the test")
-    def test_backward_per_tensor(self, device, X):
```
Comment (raghuramank100) [inline test/quantization/test_fake_quant.py:89]: This is also an op level test

File: test/quantization/test_fake_quant.py [hunk 9]
```diff
@@ -1,311 +0,0 @@
...
-            dout, X, scale, zero_point, quant_min, quant_max)
-        Y_prime.backward(dout)
-        np.testing.assert_allclose(dX.cpu(), X.grad.cpu().detach().numpy(), rtol=tolerance, atol=tolerance)
-
-    @given(device=st.sampled_from(['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']),
-           X=hu.tensor(shapes=hu.array_shapes(1, 5,),
-                       qparams=hu.qparams(dtypes=torch.quint8)))
-    # https://github.com/pytorch/pytorch/issues/30604
-    @unittest.skip("temporarily disable the test")
-    def test_numerical_consistency_per_tensor(self, device, X):
```
Comment (raghuramank100) [inline test/quantization/test_fake_quant.py:113]: Same here

**Review by raghuramank100 — APPROVED** [review]
I think we should refactor these differently as some of the tests are op level tests. Thanks for refactoring the code!

File: test/quantization/test_fake_quant.py (hunk 1 above)
Comment (jerryzh168) [reply test/quantization/test_fake_quant.py:71]: yeah will add a todo

File: test/quantization/test_fake_quant.py (hunk 2 above)
Comment (jerryzh168) [reply test/quantization/test_fake_quant.py:132]: test_fake_quant.py is renamed to test_workflow_module.py

**Pull request closed without merging**